File: api/models/item_types.py
import json

from models.base import Base
from providers import data_provider

ITEM_TYPES = []


class ItemTypes(Base):
    def __init__(self, root_path, is_debug=False):
        self.data_path = root_path + "item_types.json"
        self.load(is_debug)

    def get_item_types(self):
        return self.data

    def get_item_type(self, item_type_id):
        for x in self.data:
            if x["id"] == item_type_id:
                return x
        return None

    def add_item_type(self, item_type):
        for x in self.data:
            if x["id"] == item_type["id"]:
                return False
        item_type["created_at"] = self.get_timestamp()
        item_type["updated_at"] = self.get_timestamp()
        self.data.append(item_type)
        return True

    def update_item_type(self, item_type_id, item_type):
        if "id" in item_type:
            if item_type_id != item_type["id"]:
                return False
        item_type["updated_at"] = self.get_timestamp()
        for i in range(len(self.data)):
            if self.data[i]["id"] == item_type_id:
                self.data[i] = item_type
                return True

    def remove_item_type(self, item_type_id):
        item_type = self.get_item_type(item_type_id)
        if item_type is None: return False
        items = data_provider.fetch_item_pool().get_items_for_item_type(item_type_id)
        if len(items) != 0: return False
        for x in self.data:
            if x["id"] == item_type_id:
                self.data.remove(x)
                return True

    def load(self, is_debug):
        if is_debug:
            self.data = ITEM_TYPES
        else:
            f = open(self.data_path, "r")
            self.data = json.load(f)
            f.close()

    def save(self):
        f = open(self.data_path, "w")
        json.dump(self.data, f)
        f.close()

File: C#api/models/item_types.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using Newtonsoft.Json;
using ProvidersV2;
namespace ModelsV2;

public class ItemType
{
    public int? Id { get; set; }
    public string Name { get; set; }
    public string Description { get; set; }
    public string? Created_At { get; set; }
    public string? Updated_At { get; set; }
}


public class ItemTypes : Base
{
    private readonly string _dataPath;
    private List<ItemType> _data;

    public ItemTypes(string rootPath, bool isDebug = false)
    {
        _dataPath = Path.Combine(rootPath, "item_types.json");
        Load(isDebug);
    }

    public List<ItemType> GetItemTypes()
    {
        return _data;
    }

    public ItemType GetItemType(int itemtypeId)
    {
        return _data.Find(x => x.Id == itemtypeId);
    }

    public bool AddItemtype(ItemType itemtype)
    {
        itemtype.Id = _data.Count > 0 ? _data.Max(it => it.Id) + 1 : 1;
        if (itemtype.Created_At == null) itemtype.Created_At = GetTimestamp();
        if (itemtype.Updated_At == null) itemtype.Updated_At = GetTimestamp();
        _data.Add(itemtype);
        return true;
    }

    public bool UpdateItemtype(int itemtypeId, ItemType itemtype)
    {
        itemtype.Updated_At = GetTimestamp();
        var index = _data.FindIndex(x => x.Id == itemtypeId);
        if (index >= 0)
        {
            itemtype.Id = _data[index].Id;
            itemtype.Created_At = _data[index].Created_At;
            _data[index] = itemtype;
            return true;
        }

        return false;
    }
        public bool ReplaceItemType(int itemTypeId , ItemType newItemType)
    {
        var index = _data.FindIndex(existingItemType => existingItemType.Id == itemTypeId);
        var existingItemType = _data.FirstOrDefault(existingItemType => existingItemType.Id == itemTypeId);

        if (index < 0)
        {

            return false;

        }

        if (!string.IsNullOrEmpty(newItemType.Name)) existingItemType.Name = newItemType.Name;
        if (!string.IsNullOrEmpty(newItemType.Description)) existingItemType.Description = newItemType.Description;
        existingItemType.Updated_At = GetTimestamp();

        return true;
    }

    public bool RemoveItemtype(int itemtypeId, bool force = false)
    {
        var itemtype = GetItemType(itemtypeId);
        if (itemtype == null) return false;
        if (force) return _data.Remove(itemtype);
        var items = DataProvider.fetch_item_pool().GetItemsForItemType(itemtypeId);
        if (items.Count() != 0) return false;

        return _data.Remove(itemtype);
    }

    private void Load(bool isDebug)
    {
        if (isDebug)
        {
            _data = new List<ItemType>();
        }
        else
        {
            using (var reader = new StreamReader(_dataPath))
            {
                var json = reader.ReadToEnd();
                _data = JsonConvert.DeserializeObject<List<ItemType>>(json);
            }
        }
    }

    public void Save()
    {
        using (var writer = new StreamWriter(_dataPath))
        {
            var json = JsonConvert.SerializeObject(_data, Formatting.Indented);
            writer.Write(json);
        }
    }
}
